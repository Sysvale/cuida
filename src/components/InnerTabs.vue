<template>
	<div
		id="inner-tabs"
	>
		<div class="inner-tabs__header-container">
			<ul class="inner-tabs__header">
				<li
					v-for="(tab, index) in tabs"
					:key="`${index}-${tab.name}-tab`"
					role="presentation"
				>
					<a
						role="tab"
						href="javascript:void(0)"
						target="_self"
						class="inner-tabs__tab"
						:class="isActive(tab) ? 'inner-tabs__tab--active' : 'inner-tabs__tab--inactive'"
						@click="(event) => handleClick(event, tab)"
					>
						{{ tab.title }}
					</a>
	
					<div
						:class="isActive(tab) ? 'tab__indicator--active' : ''"
						:style="activeBorderStyle"
					/>
				</li>
			</ul>
		</div>

		<div class="inner-tabs__content">
			<div
				v-for="(tab, index) in tabs"
				:key="`${index}-${tab.name}-tab`"
				role="tabpanel"
				class="content__pane"
				:class="isActive(tab) ? 'content__pane--active' : 'content__pane--inactive'"
			>
				<!-- @slot Slot para renderização customizada do conteúdo das abas
					os slots são nomeados de acordo com os `name` das tabs, cada slot
					nomeado é exibido de acordo com a aba ativa correspondente
				-->
				<slot
					:name="getSlotName(tab)"
				/>
			</div>
		</div>
	</div>
</template>

<script>
import isEmpty from 'lodash.isempty';
import isEqual from 'lodash.isequal';

import { colorOptions, colorHexCode } from '../utils/constants/colors';

export default {
	props: {
		/**
		 * Define a lista dos itens da InnerTabs a serem mostrados.
		 * Os itens da lista devem ser objetos com `name` (para identificar o slot)
		 * e `title` (título da aba)
		 */
		tabs: {
			type: Array,
			default: () => ([]),
			required: true,
			validator: (values) => {
				const invalidValues = values.filter((value) => {
					return isEmpty(value.title) || isEmpty(value.name);
				});
				return !invalidValues.length;
			},
		},
		/**
		 * O item ativo dentre as abas
		 */
		activeTab: {
			type: Object,
			default: () => ({}),
			required: true,
		},
		/**
		 * Cor da borda que indica o item ativo.
		 * Existem algumas cores predefinidas seguindo os guias do Cuida, são elas: 
		 * `turquoise`, `green`, `blue`, `violet`, `pink`, `red`, `orange`, `amber` e `gray`.
		 */
		variant: {
			type: String,
			default: 'green',
			validator: (value) => {
				return colorOptions.includes(value);
			},
		},
	},

	data() {
		return {
			internalActiveTab: this.activeTab,
			colorOptions,
		};
	},

	computed: {
		activeBorderStyle() {
			return {
				'--indicatorColor': this.colorHexCode(this.variant),
			};
		},
	},

	watch: {
		tabs: {
			handler(newValue) {
				const filtered = newValue.filter(tab => tab.name === this.activeTab.name);
				[this.internalActiveTab] = filtered.length ? filtered : newValue;
			},
			immediate: true,
		},
	},

	methods: {
		colorHexCode,

		getSlotName(tab) {
			return tab.name;
		},

		handleClick(event, item) {
			/**
			 * Evento emitido quando a aba ativa é alterada
			* @event change
			* @type {Event}
				*/
			this.$emit('change', { event, item });
			this.internalActiveTab = item;
		},

		isActive(item) {
			return isEqual(this.internalActiveTab, item);
		},
	},
};
</script>

<style lang="scss">
@import '../assets/sass/tokens.scss';

#inner-tabs {
	.inner-tabs {
		&__header {
			display: flex;
			justify-content: flex-start;
			gap: spacer(4);
			padding: pl(0);
			margin: mb(0);
			list-style: none;
			transition: $opening;

			&-container {
				overflow-x: scroll;

				/* width */
				&::-webkit-scrollbar {
					height: 4px;
					border-radius: 8px;
				}

				/* Track */
				&::-webkit-scrollbar-track {
					background: transparent;
				}

				/* Handle */
				&::-webkit-scrollbar-thumb {
					background: $n-40;
					border-radius: 8px;
				}

				/* Handle on hover */
				&::-webkit-scrollbar-thumb:hover {
					background: $n-50;
				}
			}
		}

		&__tab {
			width: max-content;
			min-width: 140px;
			@include caption;
			font-weight: $font-weight-regular;
			overflow: hidden;
			text-overflow: ellipsis;
			display: -webkit-box;
			-webkit-line-clamp: 1;
			-webkit-box-orient: vertical;
			text-align: center;
			text-decoration: none !important;
			margin: mTRBL(0, 2, 2, 0);
			padding: pYX(2, 5);
			cursor: pointer;

			&--active {
				cursor: default;
				color: $n-800;
				font-weight: $font-weight-semibold;
			}
			
			&--active:hover {
				color: $n-800;
			}
			
			&--inactive {
				color: $n-500;
				font-weight: $font-weight-regular;
			}
			
			&--inactive:hover {
				color: $n-500;
			}
		}

		&__content {
			background-color: $n-0;
			padding: pa(5);
			border-radius: $border-radius-button;
		
			> .content__pane--inactive {
				display: none;
			}
		
			> .content__pane--active {
				display: block;
			}
		}
	}

	.tab__indicator--active {
		height: 4px;
		width: 100%;
		border-top-right-radius: 4px;
		border-top-left-radius: 4px;
		background-color: var(--indicatorColor);
		transition: all 0.3s ease-out;
	}
}
</style>
