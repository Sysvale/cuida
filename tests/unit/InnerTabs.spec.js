import { mount, createLocalVue } from '@vue/test-utils';
import lodash from 'lodash';
import InnerTabs from '../../src/components/InnerTabs.vue';

const localVue = createLocalVue();

const mockedData = [
	{ title: 'Dummy title 1', name: 'dummy-1' },
	{ title: 'Dummy title 2', name: 'dummy-2' },
	{ title: 'Dummy title 3', name: 'dummy-3' },
	{ title: 'Dummy title 4', name: 'dummy-4' },
];

test('Component is mounted properly', () => {
	window._ = lodash;
	const wrapper = mount(InnerTabs, {
		localVue,
		propsData: {
			tabs: mockedData,
			activeTab: mockedData[0],
		},
	});
	expect(wrapper).toMatchSnapshot();
});

describe('"tabs" prop is validated', () => {
	const validator = InnerTabs.props.tabs.validator;

	test('passes when "tabs" has name and title', () => {
		expect(validator(mockedData)).toBe(true);
	});

	test('throws exception when some element of "tabs" has not title', () => {
		expect(validator([{ name: 'dummy-0' }, ...mockedData])).toBe(false);
	});

	test('throws exception when some element of "tabs" has not name', () => {
		expect(validator([{ title: 'Dummy title 0' }, ...mockedData])).toBe(false);
	});
});

describe('"variant" prop is validated', () => {
	const validator = InnerTabs.props.variant.validator;

	test('passes when "variant" is a predefined color', () => {
		expect(validator('turquoise')).toBe(true);
	});

	test('throws exception when "variant" is not a predefined color', () => {
		expect(validator('dummy')).toBe(false);
	});
});

describe('Items styles test', () => {
	test('if has one tab set as active', () => {
		window._ = lodash;
		const wrapper = mount(InnerTabs, {
			localVue,
			propsData: {
				tabs: mockedData,
				activeTab: mockedData[0],
			},
		});
		expect(wrapper.findAll('.inner-tabs__tab--active').length).toBe(1);
	});
});
