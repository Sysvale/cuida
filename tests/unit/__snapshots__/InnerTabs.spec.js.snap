// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<div id="inner-tabs">
  <ul class="inner-tabs__header">
    <li role="presentation"><a role="tab" href="javascript:void(0)" target="_self" class="inner-tabs__tab inner-tabs__tab--active">
        Dummy title 1
      </a>
      <div class="tab__indicator--active" style="--indicatorColor: #2AC092;"></div>
    </li>
    <li role="presentation"><a role="tab" href="javascript:void(0)" target="_self" class="inner-tabs__tab inner-tabs__tab--inactive">
        Dummy title 2
      </a>
      <div class="" style="--indicatorColor: #2AC092;"></div>
    </li>
    <li role="presentation"><a role="tab" href="javascript:void(0)" target="_self" class="inner-tabs__tab inner-tabs__tab--inactive">
        Dummy title 3
      </a>
      <div class="" style="--indicatorColor: #2AC092;"></div>
    </li>
    <li role="presentation"><a role="tab" href="javascript:void(0)" target="_self" class="inner-tabs__tab inner-tabs__tab--inactive">
        Dummy title 4
      </a>
      <div class="" style="--indicatorColor: #2AC092;"></div>
    </li>
  </ul>
  <hr class="inner-tabs__divider">
  <div class="inner-tabs__content">
    <div role="tabpanel" class="content__pane content__pane--active"></div>
    <div role="tabpanel" class="content__pane content__pane--inactive"></div>
    <div role="tabpanel" class="content__pane content__pane--inactive"></div>
    <div role="tabpanel" class="content__pane content__pane--inactive"></div>
  </div>
</div>
`;
